core/src/main/scala/cats/data/ContT.scala

```diff
@@ -160,6 +162,38 @@ object ContT {
   def later[M[_], A, B](fn: => (B => M[A]) => M[A]): ContT[M, A, B] =
     DeferCont(() => FromFn(AndThen(fn)))
 
+  /*
+   * Limits the continuation of any inner [[shiftT]]
+   */
+  def resetT[M[_]: Monad, R, R2](contT: ContT[M, R, R]): ContT[M, R2, R] =
+    ContT.liftF(contT.eval)
+
+  /*
+   * Captures the continuation up to the nearest enclosing [[resetT]] and passes
+   * it to f
+   *
+   * {{{
+   *   for {
+   *     _ <- ContT.resetT(
+   *            for {
+   *              _ <- ContT.liftF(IO.println("1"))
+   *              _ <- ContT.shiftT{ (k: Unit => IO[Unit])
+   *                     for {
+   *                       _ <- ContT.liftF(IO.println("2"))
+   *                       _ <- ContT.liftF(k(()))
+   *                       _ <- ContT.liftF(IO.println("4"))
+   *                     }
+   *                   }
+   *              _ <- ContT.liftF(IO.println("3"))
+   *            } yield ()
+   *         )
+   *     _ <- ContT.liftF(IO.println("5"))
+   *   } yield ()
+   * }}}
+   */
+  def shiftT[M[_]: Monad, R, A](f: (A => M[R]) => ContT[M, R, R]): ContT[M, R, A] =
+    apply(cb => f(cb).eval)
```

Comment:
putting this inside a `Defer` seems more likely to be needed...

Can we add a stack-safety test here in some way?

Reply:
I don't see why `Monad` is needed here. It seems to me that only `Applicative` should be since I think `apply` requires nothing and `eval` just requires Applicative and Defer.